engine/gamesys/src/gamesys/components/comp_mesh.cpp

```diff
@@ -45,46 +46,114 @@ namespace dmGameSystem
-        dmGraphics::HTexture        m_Textures[dmRender::RenderObject::MAX_TEXTURE_COUNT];
-        dmRender::HMaterial         m_Material;
+        dmGameObject::HInstance         m_Instance;
+        Matrix4                         m_Local;
+        Matrix4                         m_World;
+        uint32_t                        m_MixedHash;
+        CompRenderConstants             m_RenderConstants;
+        MeshResource*                   m_Resource;
+        dmGameSystem::BufferResource*   m_BufferResource;
+        dmGraphics::HTexture            m_Textures[dmRender::RenderObject::MAX_TEXTURE_COUNT];
+        dmRender::HMaterial             m_Material;
 
         // If the buffer resource property has changed (ie not same buffer resource as
         // used by the mesh resource), we need to create a specific vertex declaration
         // and vertbuffer for this buffer resource instead.
         dmGraphics::HVertexDeclaration  m_VertexDeclaration;
         uint32_t                        m_ElementCount;
-        uint32_t                        m_VertSize;
+        uint16_t                        m_VertSize;
 
-        uint16_t                    m_ComponentIndex;
         /// Component enablement
-        uint8_t                     m_Enabled : 1;
+        uint8_t                         m_Enabled : 1;
         /// Added to update or not
-        uint8_t                     m_AddedToUpdate : 1;
-        uint8_t                     m_ReHash : 1;
+        uint8_t                         m_AddedToUpdate : 1;
+        uint8_t                         m_ReHash : 1;
+        uint8_t                         :5;
+    };
+
+    struct VertexBufferInfo
+    {
+        dmGraphics::HVertexBuffer m_VertexBuffer;
+        uint32_t m_RefCount;
+        uint32_t m_Version;
     };
 
     struct MeshWorld
     {
         dmResource::HFactory               m_ResourceFactory;
-        uint32_t                           m_CurrentVertexBuffer;
-        dmArray<dmGraphics::HVertexBuffer> m_VertexBuffers;
+        dmObjectPool<MeshComponent*>       m_Components;
+        dmArray<dmRender::RenderObject>    m_RenderObjects;
+        dmHashTable64<VertexBufferInfo>    m_ResourceToVertexBuffer;
+        dmArray<dmGraphics::HVertexBuffer> m_VertexBufferPool;
+        dmArray<dmGraphics::HVertexBuffer> m_VertexBufferWorld; // for meshes batched in world space
         void*                              m_WorldVertexData;
         size_t                             m_WorldVertexDataSize;
         /// Keep track of how much vertex data we have rendered so it can be
         /// reported to the profiler at end of the render dispatching.
         uint32_t                           m_RenderedVertexSize;
 
-        dmObjectPool<MeshComponent*>       m_Components;
-        dmArray<dmRender::RenderObject>    m_RenderObjects;
     };
 
+    static inline void DeallocVertexBuffer(MeshWorld* world, dmGraphics::HVertexBuffer vertex_buffer)
```

Comment:
The system here is for allowing us to share vertex buffers between mesh components that share the same resource.